src/routers/WormholeL2VoteAggregatorRouter.sol

```diff
@@ -0,0 +1,96 @@
+  /// @dev Thrown when calldata is invalid for the provided function ID.
+  error InvalidCalldata();
+
+  /// @dev Thrown when calldata provides a function ID that does not exist.
+  error FunctionDoesNotExist();
+
+  /// @dev Thrown when a function is not supported.
+  error UnsupportedFunction();
+
+  /// @param _votingToken The address of the L2 token used for voting.
+  /// @param _relayer The address of the Wormhole relayer contract.
+  /// @param _governorMetadata The address of the L2 Governor Metadata contract which holds proposal
+  /// state.
+  /// @param _l1BlockAddress The address of the contract used to fetch the L1 block number.
+  /// @param _sourceChain The Wormhole chain ID of the source chain when sending messages.
+  /// @param _targetChain The Wormhole chain ID of the target chain when sending messages.
+  constructor(
+    address _votingToken,
+    address _relayer,
+    address _governorMetadata,
+    address _l1BlockAddress,
+    uint16 _sourceChain,
+    uint16 _targetChain
+  )
+    WormholeL2VoteAggregator(
+      _votingToken,
+      _relayer,
+      _governorMetadata,
+      _l1BlockAddress,
+      _sourceChain,
+      _targetChain
+    )
+  {}
+
+  /// @dev if we remove this function solc will give a missing-receive-ether warning because we have
+  /// a payable fallback function. We cannot change the fallback function to a receive function
+  /// because receive does not have access to msg.data. In order to prevent a missing-receive-ether
+  /// warning we add a receive function and revert.
+  receive() external payable {
+    revert UnsupportedFunction();
+  }
+
+  /// @dev Wrap the default `castVote` method in a method that optimizes the calldata.
+  /// @param _msgData Optimized calldata for the `castVote` method. We restrict proposalId to a
+  /// `uint16` rather than the default `uint256`.
+  function _castVote(bytes calldata _msgData) internal {
+    if (_msgData.length != 4) revert InvalidCalldata();
+    uint16 proposalId = uint16(bytes2(_msgData[1:3])); // Supports max id of 65,535
+    uint8 support = uint8(bytes1(_msgData[3:4]));
+    castVote(proposalId, L2VoteAggregator.VoteType(support));
+  }
+
+  /// @dev Wrap the default `castVoteWithReason` method in a method that optimizes the calldata.
+  /// @param _msgData Optimized calldata for the `castVoteWithReason` method. We restrict proposalId
+  /// to a `uint16` rather than the default `uint256`.
+  function _castVoteWithReason(bytes calldata _msgData) internal {
+    uint16 proposalId = uint16(bytes2(_msgData[1:3]));
+    uint8 support = uint8(bytes1(_msgData[3:4]));
+    string calldata reason = string(_msgData[4:]);
+    castVoteWithReason(proposalId, L2VoteAggregator.VoteType(support), reason);
+  }
+
+  /// @dev Wrap the default `castVoteBySig` method in a method that optimizes the calldata.
+  /// @param _msgData Optimized calldata for the `castVoteBySig` method. We restrict proposalId to a
+  /// `uint16` rather than the default `uint256`.
```

Comment:
Does this assumption actually work? For OZ Governors, the proposal Id is a hash of the components as defined [here](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/governance/Governor.sol#L131). See this recent PoolTogether  proposal as an example of how big the IDs end up:
```
https://www.tally.xyz/gov/pooltogether/proposal/65723469450716275481895714769883720280687015505689656874523874999170247998755
```

Reply:
Ah yes I should of caught that, I can add a mapping of real ID to an auto incrementing ID that can be used in the optimized metods

Reply:
I think it could live only on this extension, but it would have to override the base method, update the mapping, then call super (or something roughly like that).

Reply:
yes, I pushed up the initial changes and plan on cleaning them up but here is what I was thinking.

https://github.com/ScopeLift/l2-flexible-voting/pull/69/commits/7723f64babe68becce3bca673cd8b7c5336bb61b